src/biogeophys/FrictionVelocityMod.F90

```diff
@@ -516,26 +557,61 @@ subroutine SetRoughnessLengthsAndForcHeightsNonLake(this, bounds, &
          urbpoi           =>    lun%urbpoi                            , & ! Input:  [logical  (:)   ] true => landunit is an urban point
          z_0_town         =>    lun%z_0_town                          , & ! Input:  [real(r8) (:)   ] momentum roughness length of urban landunit (m)
          z_d_town         =>    lun%z_d_town                          , & ! Input:  [real(r8) (:)   ] displacement height of urban landunit (m)
          forc_hgt_t       =>    atm2lnd_inst%forc_hgt_t_grc           , & ! Input:  [real(r8) (:)   ] observational height of temperature [m]
          forc_hgt_u       =>    atm2lnd_inst%forc_hgt_u_grc           , & ! Input:  [real(r8) (:)   ] observational height of wind [m]
-         forc_hgt_q       =>    atm2lnd_inst%forc_hgt_q_grc             & ! Input:  [real(r8) (:)   ] observational height of specific humidity [m]
+         forc_hgt_q       =>    atm2lnd_inst%forc_hgt_q_grc           , & ! Input:  [real(r8) (:)   ] observational height of specific humidity [m]
+         z0mg_2D          =>    this%z0mg_2D_col                        & ! Input:  [real(r8) (:)   ] 2-D field of input col roughness length over ground, momentum  [m] 
+         
          )
 
     do fc = 1, num_nolakec
        c = filter_nolakec(fc)
 
        ! Ground roughness lengths over non-lake columns (includes bare ground, ground
        ! underneath canopy, wetlands, etc.)
-       if (frac_sno(c) > 0._r8) then
-          z0mg(c) = this%zsno
-       else
-          z0mg(c) = this%zlnd
-       end if
+
+       select case (z0param_method)
+       case ('ZengWang2007')
+          if (frac_sno(c) > 0._r8) then
+             if(use_z0m_snowmelt) then
+                if ( snomelt_accum(c) < 1.e-5_r8 )then
+                    z0mg(c) = exp(-b1_param * rpi * 0.5_r8 + b4_param) * 1.e-3_r8
+                else
+                    z0mg(c) = exp(b1_param * (atan((log10(snomelt_accum(c))+0.23_r8)/0.08_r8)) + b4_param) * 1.e-3_r8
+                end if
+             else
+                z0mg(c) = this%zsno
+             end if                    
+          else
+             z0mg(c) = this%zlnd
+          end if
+       case ('Meier2022')           ! Bare ground and ice have a different value
+          l = col%landunit(c)
+          if (frac_sno(c) > 0._r8) then ! Do snow first because ice could be snow-covered
+             if(use_z0m_snowmelt) then
+                if ( snomelt_accum(c) < 1.e-5_r8 )then
+                    z0mg(c) = exp(-b1_param * rpi * 0.5_r8 + b4_param) * 1.e-3_r8
+                else
+                    z0mg(c) = exp(b1_param * (atan((log10(snomelt_accum(c))+0.23_r8)/0.08_r8)) + b4_param) * 1.e-3_r8
+                end if
+             else
+                z0mg(c) = this%zsno
+             end if                    
+          else if (lun%itype(l) == istice) then
+             z0mg(c) = this%zglc
```

Comment:
The above seems unnecessarily repetitive just to apply this `else if`. Refactoring should be bfb.